### PR TITLE
feat: VueJS + NuxtJS support (#199)

### PR DESCRIPTION
**Pull request by markflorkowski**
Add support for NuxtJS and VueJS

- [x] Create builder function for Nuxt that will scaffold API route handler
- [x] Adjust types to comply with NuxtJS type system in case of API Routes
- [x] Expose proper Request object in middleware
- [x] Create package for VueJS support
- [x] Custom ~hooks~ composables for VueJS
- [x] Button
- [x] Dropzone
- [x] Upload progress indicator in default components

Update:

Vue package with UploadButton is added. Currently, it has external dependency on [VueUse](https://github.com/vueuse/vueuse) and it's useFetch. Ideally, it should be replaced with something more lightweight

Since I've used `defineComponent` function to create UploadButton component, I had to make single top-level prop that can be passed in component config
image

So everything else can be passed into config prop object

# Example of usage in SFC:
 
image

Typesafety and autocomplete works as expected
image
image

Correct types in middleware
image

Issues present so far that has to be fixed:
- [x] Import from `@uploadthing/vue` is not working. import from `@uploadthing/vue/index` works but first option is more preferable
- [x] Can't make TailwindCSS work in NuxtJS example for some reason. It does not generate classes
- [x] Both UploadButton and UploadDropzone require `ClientOnly` to work properly. Not sure how much of an issue it is though

# Example from FE
Monosnap screencast 2023-07-06 13-26-01

### CODE REVIEW

**Comment by Mr0Bread** [comment]
Cool, thanks

**Comment by markflorkowski** [comment]
@Mr0Bread  It looks like it didn't actually publish the vue package. I think you need to add this to the `package.json`:

```json
"publishConfig": {
    "access": "public"
  },
```

cc @juliusmarminge

**Comment by juliusmarminge** [comment]
So we actually need to change the workflow abit to make sure it's versioned properly:

https://github.com/pingdotgg/uploadthing/blob/ad1c7a39f9a2fcdce69f5bc50c59758c5deb52da/.github/workflows/release-canary.yaml#L50

https://github.com/pingdotgg/uploadthing/blob/main/.github/canary-version.js

Maybe we should change it so it loops over the entire packages folder (except config)

**Comment by Mr0Bread** [comment]
@markflorkowski This branch is now part of this repo, not my fork. I'm not sure I'll be able to push into this branch

**Comment by markflorkowski** [comment]
> I'm not sure I'll be able to push into this branch

If not, fork and open PRs against this branch @Mr0Bread

File: docs/src/pages/nuxt.mdx [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+<template>
+  <main
+    class="flex min-h-screen flex-col items-center justify-center gap-16 p-24"
+  >
+    <ClientOnly fallback-tag="div" fallback="Loading...">
+      <Uploader :config="{ endpoint: 'videoAndImage', onClientUploadComplete={()
+      => { alert("Upload Completed"); }} onUploadError={(error: Error) => { //
+      Do something with the error. alert(`ERROR! ${error.message}`); }} }" />
+    </ClientOnly>
```
Comment (juliusmarminge) [inline docs/src/pages/nuxt.mdx:202]: why does this have to be in a client-only block? is it not possible to SSR this and have the file config on first paint?

File: docs/src/pages/nuxt.mdx [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+</script>
+
+<template>
+  <main
+    class="flex min-h-screen flex-col items-center justify-center gap-16 p-24"
+  >
+    <UploadButton :config="{ endpoint: 'videoAndImage',
+    onClientUploadComplete={() => { alert("Upload Completed"); }}
+    onUploadError={(error: Error) => { // Do something with the error.
+    alert(`ERROR! ${error.message}`); }} }" />
```
Comment (juliusmarminge) [inline docs/src/pages/nuxt.mdx:228]: here it isn't wrapped in ClientOnly 🤔

Comment (Mr0Bread) [reply docs/src/pages/nuxt.mdx:228]: Oh, outdated example. I made it work with SSR, so now no need in using ClientOnly

File: docs/src/pages/nuxt.mdx (hunk 1 above)
Comment (Mr0Bread) [reply docs/src/pages/nuxt.mdx:202]: This one have to be fixed. ClientOnly is no longer needed

**Review by t3dotgg — APPROVED** [review]
Good work on the dropzone in particular, that didn't look fun to do

**Comment by juliusmarminge** [comment]
Changing base for the time being as graphite really doesn't like restacking when this branch doesn't have diffs to main :thinking:

**Comment by juliusmarminge** [comment]
Current dependencies on/for this PR:
* main
  * **PR #293** <a href="https://app.graphite.dev/github/pr/pingdotgg/uploadthing/293" target="_blank">Graphite</a>  👈
    * **PR #401** <a href="https://app.graphite.dev/github/pr/pingdotgg/uploadthing/401" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/pingdotgg/uploadthing/293?utm_source=stack-comment).

File: packages/vue/src/useDropzone.ts
Comment (juliusmarminge) [inline packages/vue/src/useDropzone.ts:1]: TODO: Rewrite using https://github.com/pingdotgg/uploadthing/tree/main/packages/dropzone

**Comment by juliusmarminge** [comment]
Pretty mcuh done with the nuxt-moudle, just gotta fix the typing issue with the generics, right now it's not getting narrowed correctly:

CleanShot 2024-01-20 at 19 10 00@2x

**Comment by juliusmarminge** [comment]
Tracking the generic bug here: https://github.com/vuejs/core/issues/10167

File: packages/dropzone/src/vue.ts
Comment (juliusmarminge) [inline packages/dropzone/src/vue.ts:1]: this doesn't work rn, doesn't seem to be any interactivity with it... need a couple of vue eyes here :)

File: packages/dropzone/src/solid.tsx [hunk 3]
```diff
@@ -192,7 +186,7 @@ export function createDropzone(_props: DropzoneOptions) {
       }
     });
 
-    if (isValidQuantity(acceptedFiles, props.multiple, props.maxFiles)) {
+    if (!isValidQuantity(acceptedFiles, props.multiple, props.maxFiles)) {
```
Comment (juliusmarminge) [inline packages/dropzone/src/solid.tsx:189]: whoopsies should prob be fixed in a PR that'll get merged in the near future cc @markflorkowski

File: .vscode/settings.json [hunk 4]
```diff
@@ -5,7 +5,7 @@
   "editor.defaultFormatter": "esbenp.prettier-vscode",
   "editor.formatOnSave": true,
   "eslint.rules.customizations": [{ "rule": "*", "severity": "warn" }],
-  "explorer.excludeGitIgnore": true,
+  // "explorer.excludeGitIgnore": true,
```
Comment (markflorkowski) [inline .vscode/settings.json:8]: Can we just remove this permanently

Comment (markflorkowski) [reply .vscode/settings.json:8]: lol

image

Comment (juliusmarminge) [reply .vscode/settings.json:8]: Haha Yea we can keep it off

File: packages/nuxt/package.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  },
+  "main": "./dist/module.cjs",
+  "types": "./dist/types.d.ts",
+  "files": [
+    "dist"
+  ],
+  "scripts": {
+    "lint": "eslint src --max-warnings 0",
+    "build": "nuxt-module-build build && tsc && bun run postbuild.ts",
```
Comment (juliusmarminge) [inline packages/nuxt/package.json:20]:
```suggestion
    "build": "nuxt-module-build build",
```

File: packages/nuxt/postbuild.ts
Comment (juliusmarminge) [inline packages/nuxt/postbuild.ts:1]: 💀

File: packages/nuxt/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./.nuxt/tsconfig.json",
+  "include": ["src"],
+  "compilerOptions": {
+    "declaration": true,
+    "noEmit": false,
+    "emitDeclarationOnly": false,
+    "outDir": "dist"
+  }
```
Comment (juliusmarminge) [inline packages/nuxt/tsconfig.json:9]:
```suggestion
```

**Review by juliusmarminge — COMMENTED** [review]
Fun bundle hacks :)

**Pull request merged**